extensions/flight-sql/src/main/java/io/deephaven/server/flightsql/FlightSqlResolver.java

```diff
@@ -0,0 +1,1710 @@
+                if (totalRecords != -1) {
+                    if (table.isRefreshing()) {
+                        throw new IllegalStateException(
+                                "TicketHandler implementation error; should only override totalRecords for non-refreshing tables");
+                    }
+                    if (table.size() != totalRecords) {
+                        throw new IllegalStateException(
+                                "Ticket handler implementation error; totalRecords does not match the table size");
+                    }
+                }
+                return table;
+            }
+        }
+    }
+
+    private static final class UnsupportedCommand<T> implements CommandHandler<T>, TicketHandler {
+        private final Descriptor descriptor;
+
+        UnsupportedCommand(Descriptor descriptor) {
+            this.descriptor = Objects.requireNonNull(descriptor);
+        }
+
+        @Override
+        public TicketHandler initialize(T command) {
+            return this;
+        }
+
+        @Override
+        public boolean isOwner(SessionState session) {
+            return true;
+        }
+
+        @Override
+        public FlightInfo getInfo(FlightDescriptor descriptor) {
+            throw error(Code.UNIMPLEMENTED,
+                    String.format("command '%s' is unimplemented", this.descriptor.getFullName()));
+        }
+
+        @Override
+        public Table resolve() {
+            throw error(Code.INVALID_ARGUMENT, String.format(
+                    "client is misbehaving, should use getInfo for command '%s'", this.descriptor.getFullName()));
+        }
+    }
+
+    abstract class QueryBase<C> implements CommandHandler<C>, TicketHandlerReleasable {
+        private final ByteString handleId;
+        protected final SessionState session;
+
+        private boolean initialized;
+        private boolean resolved;
+        private Table table;
+
+        QueryBase(SessionState session) {
+            this.handleId = ByteString.copyFromUtf8(UUID.randomUUID().toString());
+            this.session = Objects.requireNonNull(session);
+            queries.put(handleId, this);
+        }
+
+        public ByteString handleId() {
+            return handleId;
+        }
+
+        @Override
+        public final TicketHandlerReleasable initialize(C command) {
+            try {
+                return initializeImpl(command);
+            } catch (Throwable t) {
+                release();
+                throw t;
+            }
+        }
+
+        private synchronized QueryBase<C> initializeImpl(C command) {
+            if (initialized) {
+                throw new IllegalStateException("initialize on Query should only be called once");
+            }
+            initialized = true;
+            execute(command);
+            if (table == null) {
+                throw new IllegalStateException(
+                        "QueryBase implementation has a bug, should have set table");
```

Comment:
arguable QueryBase should be getClass(), as it was the impl's execute() that failed to call executeImpl?

Also do we want to use Assert for these sorts of "we broke our own contract" assertions?

Reply:
Updated to Assert.